File: backend/app/api/v1/clients.py
"""Client endpoints."""

from typing import Annotated
import math

from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.api.deps import get_current_active_user
from app.models import User
from app.services.client_service import ClientService
from app.schemas.client import ClientCreate, ClientUpdate, ClientResponse

router = APIRouter()


@router.get("")
async def list_clients(
    current_user: Annotated[User, Depends(get_current_active_user)],
    db: AsyncSession = Depends(get_db),
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=100),
    search: str = Query(None),
):
    """List all clients for the current agency."""
    if not current_user.agency_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="User is not associated with an agency",
        )

    clients, total = await ClientService.list_by_agency(
        db=db,
        agency_id=current_user.agency_id,
        page=page,
        per_page=per_page,
        search=search,
    )

    return {
        "items": clients,
        "total": total,
        "page": page,
        "per_page": per_page,
        "pages": math.ceil(total / per_page) if total > 0 else 0,
    }


@router.post("", response_model=ClientResponse, status_code=status.HTTP_201_CREATED)
async def create_client(
    client_data: ClientCreate,
    current_user: Annotated[User, Depends(get_current_active_user)],
    db: AsyncSession = Depends(get_db),
):
    """Create a new client."""
    if not current_user.agency_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="User is not associated with an agency",
        )

    client = await ClientService.create(
        db=db,
        agency_id=current_user.agency_id,
        name=client_data.name,
        industry=client_data.industry,
        website=client_data.website,
        location=client_data.location,
        description=client_data.description,
        services=client_data.services,
        keywords=client_data.keywords,
        tone=client_data.tone,
    )
    return client


@router.get("/{client_id}", response_model=ClientResponse)
async def get_client(
    client_id: str,
    current_user: Annotated[User, Depends(get_current_active_user)],
    db: AsyncSession = Depends(get_db),
):
    """Get a specific client."""
    if not current_user.agency_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="User is not associated with an agency",
        )

    client = await ClientService.get_by_id(db, client_id, current_user.agency_id)
    if not client:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Client not found",
        )
    return client


@router.patch("/{client_id}", response_model=ClientResponse)
async def update_client(
    client_id: str,
    update_data: ClientUpdate,
    current_user: Annotated[User, Depends(get_current_active_user)],
    db: AsyncSession = Depends(get_db),
):
    """Update a client."""
    if not current_user.agency_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="User is not associated with an agency",
        )

    client = await ClientService.get_by_id(db, client_id, current_user.agency_id)
    if not client:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Client not found",
        )

    update_dict = update_data.model_dump(exclude_unset=True)
    client = await ClientService.update(db, client, **update_dict)
    return client


@router.delete("/{client_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_client(
    client_id: str,
    current_user: Annotated[User, Depends(get_current_active_user)],
    db: AsyncSession = Depends(get_db),
):
    """Delete a client."""
    if not current_user.agency_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="User is not associated with an agency",
        )

    client = await ClientService.get_by_id(db, client_id, current_user.agency_id)
    if not client:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Client not found",
        )

    await ClientService.delete(db, client)

File: frontend/src/lib/api/clients.ts
import { apiClient } from "./client";

export interface Client {
  id: string;
  name: string;
  slug: string;
  industry: string | null;
  website: string | null;
  location: string | null;
  description: string | null;
  logo_url: string | null;
  services: string[] | null;
  keywords: string[] | null;
  tone: string;
  is_active: boolean;
  agency_id: string;
  created_at: string;
  updated_at: string;
}

export interface ClientCreate {
  name: string;
  industry?: string;
  website?: string;
  location?: string;
  description?: string;
  services?: string[];
  keywords?: string[];
  tone?: string;
}

export interface ClientUpdate {
  name?: string;
  industry?: string;
  website?: string;
  location?: string;
  description?: string;
  logo_url?: string;
  services?: string[];
  keywords?: string[];
  tone?: string;
}

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  per_page: number;
  pages: number;
}

export const clientsApi = {
  list: async (
    page = 1,
    perPage = 20,
    search?: string
  ): Promise<PaginatedResponse<Client>> => {
    const params = new URLSearchParams({
      page: page.toString(),
      per_page: perPage.toString(),
    });
    if (search) params.append("search", search);

    const response = await apiClient.get(`/clients?${params}`);
    return response.data;
  },

  get: async (id: string): Promise<Client> => {
    const response = await apiClient.get(`/clients/${id}`);
    return response.data;
  },

  create: async (data: ClientCreate): Promise<Client> => {
    const response = await apiClient.post("/clients", data);
    return response.data;
  },

  update: async (id: string, data: ClientUpdate): Promise<Client> => {
    const response = await apiClient.patch(`/clients/${id}`, data);
    return response.data;
  },

  delete: async (id: string): Promise<void> => {
    await apiClient.delete(`/clients/${id}`);
  },
};
